### PR TITLE
fix: Stream PostgreSQL data in

### LINKED ISSUE #1570: PostgreSQL Data Connector to Support Streaming
REMEMBER, BE **SMART**!

S: Specific
M: Measurable
A: Achievable
R: Relevant
T: Time-Bound

## Goal-State/What/Result

PostgreSQL data connector is implemented to support streaming to accelerators.

## Why/Purpose

Clear and concise answer to why this should be done now.

## By When

A target date for done-done completion of the entire enhancement.

**Issue/Spec written and reviewed:** <Target Date>
**Done-Done:** <Target Date>

## Done-Done

- [ ] [First Principles Driven](https://github.com/spiceai/spiceai/blob/trunk/docs/PRINCIPLES.md)
- [ ] The Algorithm
- [ ] PM/Design Review
- [ ] UX Review
- [ ] Threat Model / Security Review
- [ ] Tests
- [ ] Documentation
- [ ] Samples/Quickstarts/Tutorials
- [ ] Release Notes

## The Algorithm

- [ ] Every requirement questioned?
- [ ] Delete (Scope) any part you can.
- [ ] Simplify.
- [ ] Break down into smaller iterations/milestones.
- [ ] Opportunities for automation.

## Spec

Basic specification of the enhancement.

## How/MAP

A clear and concise plan of how this should be implemented.

### Implementation (issues)

- 
-

### PR DESCRIPTION
**Pull request by peasee**
## ❓ What type of PR is this?

- [ ] ♻️ Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## 🗣 Description

Converts PostgreSQL queries from using `.query()` to using `.query_raw()` so we can stream rows out in chunks instead of collecting all of the rows all at once in memory.

This should avoid high memory usage (and OOM errors) when loading large PostgreSQL data sources.

## 🔨 Related Issues

* Closes #1570

### CODE REVIEW

File: crates/db_connection_pool/src/dbconnection/postgresconn.rs [hunk 1]
```diff
@@ -136,11 +143,58 @@ impl<'a>
...
+
+        let first_chunk = stream.next().await;
+        if first_chunk.is_none() {
+            return Ok(Box::pin(RecordBatchStreamAdapter::new(
+                Arc::new(Schema::empty()),
+                stream::empty(),
+            )));
+        }
+
+        let first_chunk =
```
Comment (Jeadie) [inline crates/db_connection_pool/src/dbconnection/postgresconn.rs]: You can add line-level ignores for specific clippy rules.

File: crates/db_connection_pool/src/dbconnection/postgresconn.rs [hunk 2]
```diff
@@ -136,11 +143,58 @@ impl<'a>
...
+        let mut stream = streamable.chunks(4000).boxed().map(|rows| {
+            let rows = rows
+                .into_iter()
+                .collect::<std::result::Result<Vec<_>, _>>()
+                .context(QuerySnafu)?;
+            let rec = rows_to_arrow(rows.as_slice()).context(ConversionSnafu)?;
+            Ok::<_, PostgresError>(rec)
+        });
+
+        let first_chunk = stream.next().await;
```
Comment (Jeadie) [inline crates/db_connection_pool/src/dbconnection/postgresconn.rs]:
```rust
        let Some(first_chunk) = match stream.next().await else {
            return Ok(Box::pin(RecordBatchStreamAdapter::new(
                Arc::new(Schema::empty()),
                stream::empty(),
            )));
        };
```
Might resolve the below `.unwrap_or(`

File: crates/db_connection_pool/src/dbconnection/postgresconn.rs [hunk 3]
```diff
@@ -136,11 +143,58 @@ impl<'a>
...
+        // if they haven't we should use .copy_out instead, because it should be much faster
+        let streamable = self
+            .conn
+            .query_raw(sql, params.iter().copied()) // use .query_raw to get access to the underlying RowStream
+            .await
+            .context(QuerySnafu)?;
+
+        // chunk the stream into groups of 4k rows
+        //
+        // TODO: make chunk count configurable?
```
Comment (Jeadie) [inline crates/db_connection_pool/src/dbconnection/postgresconn.rs]: I think this is important enough to consider in this PR

File: crates/db_connection_pool/src/dbconnection/postgresconn.rs [hunk 4]
```diff
@@ -136,14 +143,76 @@ impl<'a>
...
+
+const BYTES_PER_KILOBYTE: u64 = 1_024;
+const BYTES_PER_MEGABYTE: u64 = 1_024 * BYTES_PER_KILOBYTE;
+const ROWS_PER_MB: u64 = 1; // 1024 rows per GB
+
+fn determine_chunk_size() -> u64 {
+    let mut sys = System::new_all();
+    sys.refresh_all();
+
+    let mem = sys.available_memory() / BYTES_PER_MEGABYTE; // available memory in MB
```
Comment (phillipleblanc) [inline crates/db_connection_pool/src/dbconnection/postgresconn.rs]: This is a relatively expensive operation: https://docs.rs/sysinfo/latest/sysinfo/#good-practice--performance-tips

We only want to fetch the data we actually use.

File: crates/runtime/tests/postgres/mod.rs [hunk 5]
```diff
@@ -22,13 +22,15 @@ use arrow::{
...
 use datafusion::execution::context::SessionContext;
+use serial_test::serial;
 use sql_provider_datafusion::SqlTable;
 
 use crate::init_tracing;
 
 mod common;
 
 #[tokio::test]
+#[serial]
```
Comment (phillipleblanc) [inline crates/runtime/tests/postgres/mod.rs]: I would prefer to keep these tests running in parallel - I think that means we need to customize the Docker container runner to specify a port (or maybe pick a random port) so we can run multiple instances at once.

**Review by phillipleblanc — DISMISSED** [review]

**Review by y-f-u — APPROVED** [review]

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**